reference/src/layout/scalars.md

```diff
@@ -37,15 +37,15 @@ The `isize` and `usize` types are pointer-sized signed and unsigned integers.
 They have the same layout as the [pointer types] for which the pointee is
 `Sized`, and are layout compatible with C's `uintptr_t` and `intptr_t` types.
 
+> **Note**: As a result, Rust's `usize` is **not** necessarily layout
+> compatible with C's `size_t`, since `uintptr_t` may be different from
+> `size_t`.
```

Comment:
An example of such an architecture might be CHERI, where `uintptr_t` is 128-bit, but `size_t` is 64-bit. Maybe @parched knows more about that. 

From my experience interacting with `libc` users, a lot of them write code under the assumption that `usize == uintptr_t == size_t`, which is indeed true for AFAIK all architectures that Rust supports today.

Reply:
Earlier this year I made some effort to get Rust (just `libcore`) to compile for CHERI. In many ways it "just works", although you have to fudge the 64-bit index size (generated from `ptrtoint`) with 128-bit `usize`, for actual indexing. Then there are the problems like 128-bit indices into `memcpy` intrinsics.

Aside: proper support absolutely requires an index type as well as a pointer size type. I think the latter could be implicit (set only in the data layout string and `target_ptr_width`), although someone will probably cook up some Unsafe code to prove me wrong.

I write this here because there really hasn't been any appropriate place to have this discussion yet.

Reply:
@nw0 you probably want to open a rust-lang/rust issue about this, since it impacts to many parts of the toolchain, and move the discussion there. We currently use `usize` for indexing and for converting pointers to ints (`ptr as usize`), so I'm not sure how your "implicit" pointer size type is compatible with both.